### ISSUE TITLE
Consider behavior change for `OptionT<T>`

### ISSUE THREAD

**Issue by iwburns** [iwburns/nullshield#50]
I have just realized that our `OptionT<T>` as it currently behaves is not - strictly speaking - a proper Functor.

Before I dive into why this is the case, I'd like to say that I'm not 100% sure if we care about this.  I can see arguments for both sides here and I'd like your opinions on how to move forward here.  Basically this comes down to one question:

------------
Should `OptionT<T>` itself prevent `null`/`undefined` from existing internally?

Or should it _only_ provide an environment in which one _can_ avoid `null`/`undefined` in a more convenient way (but not actually step in and prevent those values from coming up)?

--------------

The reason `OptionT<T>` is not a proper Functor is because it does not adhere to the "identity" rule defined for Functors which states:
`u.map(a => a)` is equivalent to `u` itself (see more info [here](https://github.com/fantasyland/fantasy-land#functor)).

The reason `OptionT<T>` does not adhere to this rule is because it checks internally what the return value is of the given function after the map process.  If the value is `null` or `undefined` it returns a `None`, otherwise, it returns a `Some<T>`.

To properly adhere to the "identity" rule, in this scenario:
```
const val = OptionT.some(5).map(a => null);
```
`val` must be `Some(null)` instead of a `None`.

If we strictly adhere to the definition of a Functor, we must allow values like `Some(null)` (and _I believe_ `Some(undefined)` as well) to be created with this library.

In addition, `OptionT<T>` is not currently a proper "Applicative" either because its `of` method checks the values that are provided and prevents a `Some(null)` or `Some(undefined)` from being created. (see more [here](https://github.com/fantasyland/fantasy-land#applicative)).

We could easily get around this and make a `fromNullable` function (which is not a part of any of these algebraic laws) to create "safe" (not `null`/`undefined`) values if we'd like.

In the end though the question remains: do we care about adhering strictly to the algebraic laws that define what Functor, Monad, Applicative, etc. all mean, or are we happy with being "close enough" for our use-case?

* On one hand, adhering to these algebraic laws allows us to depend on things like composability, reflexivity, transitivity, associativity, etc. across chains of operations dealing with `OptionT<T>`s.
* On the other hand, adhering to these laws forces us to allow some odd behavior that may be very confusing for new-comers.

[More discussion](https://github.com/ramda/ramda-fantasy/issues/163#issuecomment-322577107) about this in relation to ramda.

What do you guys think here?  There are probably quite a few implications of making a change like this, so I don't really want to make any moves lightly here.

cc
@ryanguill @brycehipp

**Comment by ryanguill** [comment]
So i've been reading through those issues and i've changed my mind a bit as I went.  My first reaction is, if the question/problem comes up that often then maybe its the implementation thats wrong...

(side note, I'm about to write a bunch of stuff as I think through it, im telling myself these things as much as anything and asking for validation).

But the core of the issue is that the `Maybe` monad is there to represent the union of "I have a value" and "I dont have a value".  That unfortunately is distinct from "I have a non-null value" and "I have a null value" - many times we mean the former when we have the latter, but not always.

Think of it like a nullable column in a database - if the column is non-nullable, a null value means `None` - if the column _is_ nullable, a null value may be a `Some` - a valid value for that thing.  In that case, `None` may mean "I dont have a value yet" and when you do, the answer may be "null".

Unfortunately, this is completely dependent on the circumstances you find yourself in - just like in a given database table the column may or may not be nullable. 

---

I think where I come down on this is that our use of `of` itself is wrong - we shouldnt be inspecting its value.  I'm pretty sure im the one that asked for this - and I still think its very useful.  I'd wager _most_ of the time a null value means `none`.  So I think like you mention, create a different function that will do that check that isnt `of` - maybe `from` - or something more explicit.  That doesnt help you when you return null from a map though - which again may be completely legitimate.  So we'll just have to get used to `.map(somethingThatMayReturnNull).filter(x => x != null)`.  Maybe that filter can be a standard thing - `.convertNullOrUndefinedToNone()` or something catchier.

**Comment by iwburns** [comment]
I think you've ended up pretty much exactly where I am.

The crazy thing is that (in my mind) it's all really based on the fact that `null` _is_ a value in `JS` and people use it _on purpose_ for _real stuff_ (which isn't a bad thing necessarily. even if you disagree with `null` itself being a thing, there's nothing we can do about it), so we simply have to support `Some(null)` if we wan't to be "totally correct" or whatever you want to call it.

With that in mind I agree about `of`.  We just need to make it not check the value passed in.  The only hang-up I have here is `undefined`.  Is checking whether or not we're storing `undefined` okay?  Technically no, but also, if you did something like:
```
const val = OptionT.of(undefined).map(x => x.prop).unwrapOr('default');
```
That `map` is going to fail.  Then again, it would fail if you had a `null` in there too.  I'm slightly torn here, but I guess it is what it is?

And about doing a `.map(somethingThatMayReturnNull)` like you said above.  I believe the common argument for how to get around that is that you should instead use `.flatMap(somethingThatReturnsAnOptionT)`.  However, that is a little unfortunate because it puts more burden on the library user to "do things properly".  I guess that's always the case, but for some reason this feels a bit more tedious.

**Comment by ryanguill** [comment]
```const val = OptionT.of(undefined).map(x => x.prop).unwrapOr('default');```

this fails because you are trying to do an operation that the type doesnt support.  This fails in the exact same way:

``` const val = OptionT.of([]).map(x => x.hasOwnProperty("foo")).unwrapOr('default');```

If you are using typescript, you'll still have to type things as `type?` or `type | undefined` to be able to put `undefined` in your option.

Youre right about `flatMap`, thats really the right answer there. I always think about it flattening a multi-dimensional collection, but it filters nulls properly too.

**Comment by iwburns** [comment]
>If you are using typescript[...]

Very true, and I think in TS having a `Some()` contain a `null` or `undefined` is probably okay because we do have the type system to help us.  In fact, I wonder what TS would say if you tried to `.map` on a `Some(undefined)` and actually reference the argument passed into the mapping function.  I'll have to play around with that and see.

However, I was originally thinking of the JS world where we don't have the same type system to help us as much.  Basically, if you _think_ your `OptionT` contains an object, but it actually contains `undefined` / `null`, then you can't do much with it safely (in JS).  I'm not sure how often that situation is going to come up, but I could see it being a problem.

**Comment by ryanguill** [comment]
> Basically, if you think your OptionT contains an object, but it actually contains undefined / null, then you can't do much with it safely (in JS).

right, but I guess the point is that isnt what this library is going to help you with - at least not if we are going to follow these laws.

My only question now is, is `of` just another name for `some()`?  Do they do anything different at all? If not, whats the purpose of `of`? Even though its shorter, it seems like it would just promote confusion about what its actually doing.

**Comment by iwburns** [comment]
>right, but I guess the point is that isnt what this library is going to help you with[...]

You're definitely right, and that leads me to the worst question of all: is the project's name still representative of what we are doing here?  I'm thinking it's not (assuming we make these changes), but I hate coming up with names.  You think we should change it?  That seems like a huge change to me, but I don't want people to use this assuming it will completely save them from `null`/`undefined` errors when it won't.

>My only question now is[...]

I believe you are correct. After these changes, `OptionT.of(x)` will just create a `Some(x)`, regardless of the type of `x` or if it even exists.  This is exactly what `OptionT.some(x)` will do as well.  We'll have to pull out all of the logic around passing in `null`/`undefined` for `some` (and `none` as well I guess), but otherwise `of` can go away in favor of the `fromNullable` function which will behave exactly as `of` does right now.

I think some of your confusion over `of` in this case might come from the fact that the fantasy-land spec defines an `of` method which has to have certain properties.  In our case, it is equivalent to both `some` and `none` which means `some` and `none` must adhere to whatever properties `of` has in the spec.

**Comment by iwburns** [comment]
So, if we _do_ decide to rename `nullshield` (not saying we should at this point) I'd like to toss `joinery` out there for consideration.

First, `join` is another name for the `flatMap` operation.  Second, `joinery` (as-in woodworking) is a bit like _composing_ two pieces of wood into a single piece (maybe a bit of a stretch there).  Since Monads are all about composition I figured it would be appropriate.

**Comment by iwburns** [comment]
At this point I think we should move forward with these changes, so I'm going to provide a summary below listing all of the things I know of that we'll need to change to get `Optional` working "properly".

---

|Function|Action|Notes|
|---|---|---|
|`of`|_remove_||
|`some`|_keep_|Need to remove checks on `val` argument.|
|`none`|_keep_|Need to remove checks on `val` argument and remove `val` entirely.|
|`map`|_keep_|Remove checks on value returned by `func` argument.|
|`fromNullable`|_add_|Will be how we allow users to prevent `null`/`undefined` inside `Optional`s if they would like.|
|`ap`|_add_|While not strictly necessary to have for this library, this is required for `Optional` to be a "true" Monad (behavior details below).|

`ap` has this signature:
```
class Some<T> {
  //...

  ap<U>(opt: Optional<(val: T) => U>): Optional<U>

  //...
}
```
In other words, it's like `map` except that the argument passed is an `Optional` containing a function instead of the function itself.

Also, I'm not 100% sold on this name.  `ap` is the "standard" name for this, but I think `apply` reads better.  However, `apply` has a special / known meaning in JS / TS.  Do you guys have any ideas on a better name here?

---

With all of the above done, I _believe_ we should be in the clear.  Do either of you have any objections to these changes?  Any other questions we should resolve before moving forward?

cc: @ryanguill @brycehipp

**Comment by ryanguill** [comment]
> So, if we do decide to rename nullshield (not saying we should at this point) I'd like to toss joinery out there for consideration.

I think I am in favor of changing the name, but im not crazy about that one. That said, I don't have any alternatives to offer.

> of - remove

I think we keep `of` and just have it work like some.  

> `map` ...

just to confirm, `flatMap` does/will do the same as `map` but will unwrap a returned `Optional` _and_ will turn null/undefined into a `None`?

> `ap`

I think we stick with the name `ap`, but if you want to provide aliases for it that works too.

Can you give a couple real world examples of `ap`?

**Comment by iwburns** [comment]
>I think I am in favor of changing the name[...]

Fair enough, I'm trying to come up with other options as well, but you guys know how hard naming things is.  Any and all suggestions are welcome.

>I think we keep `of`[...]

Is there any particular reason you'd like to do this?  Do you just like the way `Optional.of(42)` reads over `Optional.some(42)`?

>just to confirm[...]

You're partially correct here, let me explain a bit better:
- `map` will apply the provided function to the value inside the given `Optional` and will then wrap up the result in a new `Optional` before returning it.
    ```
    const two = Optional.some(1).map(x => x * 2);
    console.log(two.forceUnwrap() === 2); // true
    ```
- `flatMap` will apply the provided function to the value inside the given `Optional` _but_ will **not** wrap up the result in a new `Optional` before returning it.
    ```
    const two = Optional.some(1).map(x => Optional.some(x * 2));
    console.log(two.forceUnwrap() === 2); // true
    ```
`flatMap` itself won't actually turn null/undef into a `None`.  I think my statement before where I said:
>I believe the common argument for how to get around that is that you should instead use .flatMap(somethingThatReturnsAnOptionT)

may have been a bit confusing / misleading.

What I meant was that instead of `map`-ing with a function like this:
```
function mapping(x) {
  return x < 5 ? null : x * 2;
}
```
you should `flatMap` with a function like this:
```
function mapping(x) {
  return x < 5 ? Optional.none() : Optional.some(x * 2);
}
```

>Can you give a couple real world examples of `ap`?

This one is definitely less common, at least in the use-cases that I can think of.  Here's a swing at a "simple" plausible one that could be generalized to more complex and possibly more useful use-cases:
```
const makeAdder = x => y => x + y; // curried add

const plusOne = Optional.some(1).map(makeAdder);
// plusOne is an Optional containing a function

const three = Optional.some(2).ap(plusOne);
// here we apply the function in plusOne to `2` and wrap the result up in an `Optional`
```
Or this slightly different one:
```
const makeDevider = x => {
  if (x === 0) {
    return Optional.none();
  }
  const divider = y => y / x;
  return Optional.some(divider);
};
// if we can safely divide by the given value, return a Some containing a divider function.
// otherwise return a none.

const divide1 = makeDivider(1);
const two = Optional.some(2).ap(divide1);
// apply the divide (by 1) function to the 2 and wrap up the result

const divide0 = makeDivider(0);
const nope = Optional.some(2).ap(divide0);
// return a `None` because divide (by 0) gave us a `None`
```

**Comment by iwburns** [comment]
@ryanguill @brycehipp I know it has been a while, but do either of you have any thoughts on this?

Personally I'm a bit annoyed with the changes that we're proposing here, but I _do_ think they're necessary.

At this point I've wrapped my head around all of this as much as I can, but that doesn't mean there isn't room for more analysis / view points / considerations / etc. that I haven't thought of.

**Comment by iwburns** [comment]
@ryanguill @brycehipp following up here again.

I want to make sure this is a good move before we make it, so I don't want to rush the process here; but also want to make sure we keep moving forward and don't get stuck floundering on this issue.

#### Proposed Changes ####
|Function|Action|Notes|
|---|---|---|
|`of`|_remove_ / _keep_|Either is fine, as long as it doesn't check the `val` passed in.|
|`some`|_keep_|Need to remove checks on `val` argument.|
|`none`|_keep_|Need to remove checks on `val` argument and remove `val` entirely.|
|`map`|_keep_|Remove checks on value returned by `func` argument.|
|`fromNullable`|_add_|Will be how we allow users to prevent `null`/`undefined` inside `Optional`s if they would like.|
|`ap`|_add_|While not strictly necessary to have for this library, this is required for `Optional` to be a "true" Monad (behavior details above in convo).|

#### Reasoning ####
The idea is that `null` and `undefined` are "real" values in JS and TS; so we _technically_ need to support them inside `Optional<T>`.  It's up to the user whether they want to build a `Some(null)` value with this library or not.  In some cases that's a perfectly valid thing to do, in others it's not.  We probably shouldn't prevent that from happening.

I don't want to move forward if you guys think this "breaks" the library, but I'd also like to get this resolved because I'm pretty sure this is the last open question before we can make the move to v1.0.

Do you guys have concerns / questions here?  Should we move forward or scratch this proposal?

**Comment by iwburns** [comment]
@ryanguill @brycehipp 

I understand people are busy (myself included) with personal stuff and don't always have time to set aside for this project, but I also want to keep this moving when and where I can.

With that said, I hope you guys don't mind; but I'm going to move forward with this process as I believe it is the right move.

I'll be working on this over the next few days, but I don't expect it to take too long to test and implement.  I'd love your feedback if you guys have time to give it, but I completely understand if you don't.
